File: trunk/urlsigning/urlsigner.py
#!/usr/bin/python
# coding: utf8

import sys
import hashlib
import urllib
import hmac
import base64
import urlparse

print("")
print("URL Signer 1.0")
print("")

# Collect the URL input string (which must alrady be URL-encoded)
# from the user and parse it into its collective components
# For testing purposes, if no URL is entered, we use a default
inputStr = raw_input('URL to Sign (Must be URL-Encoded) --> ')
if not inputStr:
  inputStr = "YOUR_URL_TO_SIGN"

print("URL To Sign: " + inputStr)
url = urlparse.urlparse(inputStr)

# Collect the private key. You may want to store this
# in a private file.
privateKey = raw_input('Private Key ---> ')
if not privateKey:
  privateKey = "YOUR_PRIVATE_KEY"
  
print("Private Key: " + privateKey)

#We only need to sign the path+query part of the string
urlToSign = url.path + "?" + url.query
print("")
print("Original Path + Query: " + urlToSign)

# Decode the private key into its binary format
# We need toe decode the URL-encoded private key
decodedKey = base64.urlsafe_b64decode(privateKey)

# Create a signature using the private key and the URL-encoded
# string using HMAC SHA1. This signature will be binary.
signature = hmac.new(decodedKey, urlToSign, hashlib.sha1)

# Encode the binary signature into base64 for use within a URL
encodedSignature = base64.urlsafe_b64encode(signature.digest())
print("")
print("B64 Signature: " + encodedSignature)
originalUrl = url.scheme + "://" + url.netloc + url.path + "?" + url.query
print("")
print("Full URL: " + originalUrl + "&signature=" + encodedSignature)

File: trunk/urlsigning/urlsigner.pl
#!/usr/bin/perl
#
# Copyright (c) 2010 iGLASS Networks (http://www.iglass.net). All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
# The program contained herein is provided to you "AS IS" without any
# warranties of any kind.
#

# Required Perl Libraries.
use Digest::HMAC_SHA1;
use MIME::Base64;
use URI::URL;

use 5.006;
use strict;
use warnings;

sub decode_urlsafe_base64 {
  my($content) = @_;

  # replace '+' by '-', and '/' by '_'
  $content =~ tr/-/\+/;
  $content =~ tr/_/\//;

  return decode_base64($content);
}

sub encode_urlsafe{
  my($content) = @_;
  $content =~ tr/\+/\-/;
  $content =~ tr/\//\_/;

  return $content;

}

sub sign_url {
  my($url) = @_;

  # Private Key - Replace with your private key.
  my $KEY_STRING = 'vNIXE0xscrmjlyV-12Nj_BvUPaw=';

  # Decode the private key into original binary format.
  my $key = decode_urlsafe_base64($KEY_STRING);

  # Take only the query part
  my $parsed_url = URI::URL->new($url);
  my $url_to_sign = $parsed_url->path_query;

  # Create HMAC-SHA1 Signature
  my $digest = Digest::HMAC_SHA1->new($key);
  $digest->add($url_to_sign);

  # Encode binary signature.
  my $signature = $digest->b64digest;

  # Make Signature URL safe
  $signature = encode_urlsafe($signature);

  return ( $parsed_url->scheme .'://' .  $parsed_url->host .
          $url_to_sign .'&signature=' . $signature )
}


#
# The following is an example from Google w/ result
# See http://code.google.com/apis/maps/documentation/premier/guide.html#SignatureExamples
#

# Propertly encoded URL - Replace with your URL to be signed.
my $url = "http://maps.google.com/maps/api/geocode/json?address=New+York&sensor=false&client=clientID";

my $signed_url = sign_url($url);

print $signed_url . "\n";
